parsec/mca/device/device.c

```diff
@@ -223,6 +228,8 @@ int parsec_select_best_device( parsec_task_t* this_task ) {
             /* Skip the device if it is disabled for the taskpool */
             if(!(tp->devices_index_mask & (1 << dev_index))) continue;
             dev = parsec_mca_device_get(dev_index);
+            /* Is automatic CPU / GPU load balancing enabled? */
+            if(dev->type == PARSEC_DEV_CPU && !parsec_device_load_balance_allow_cpu) continue;
```

Comment:
This is a terrible approach. You keep the CPU chore and evaluate it for every single task just such that at the end you drop the task because `parsec_device_load_balance_allow_cpu` does not allow rebalancing ?

If there is no rebalance remove the CPU chore, such that you never have to test it again. Or if you don't want to do that, move this test up beofre calling the `eval` function and update the `valid_types` accordingly.

Reply:
1. An `eval` function, user-written or not, should never select an unallowed device. Moreover, without `parsec_device_load_balance_allow_cpu` CPU should not even be an option.
2. This is not what I suggested. I suggested to move the test before the `eval` is called, such that without `parsec_device_load_balance_allow_cpu` the `eval` for CPU will never be called.
3. It would be nice to have all devices similar, no special treatment for CPU
4. What is `advise_on_device` in this context ?